guppylang/cfg/builder.py

```diff
@@ -270,7 +271,9 @@ def visit_NamedExpr(self, node: ast.NamedExpr) -> ast.Name:
         # assignment statement and replace the expression with `x`.
         if not isinstance(node.target, ast.Name):
             raise InternalGuppyError(f"Unexpected assign target: {node.target}")
-        assign = ast.Assign(targets=[node.target], value=self.visit(node.value))
+        assign = ast.Assign(
+            targets=[copy.deepcopy(node.target)], value=self.visit(node.value)
+        )
```

Comment:
Is this a drive-by?

Reply:
No this is required now that we annotate the LHS of assignments with types

https://github.com/CQCL/guppylang/blob/f0dec41f8e722cbbbf2a9f98a7980a85f67c9a56/guppylang/checker/stmt_checker.py#L49-L56

If we don't make a copy, the type checker will see that a type has been assigned when the variable is used the next time and won't bother checking again which leads incorrect behaviour